---
title: "Preprocess 10x Genomics scATAC-seq data"
author: 
 - Lucas Graybuck
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    code_folding: hide
    df_print: paged
    self_contained: true
params:
  in_fra: NULL
  in_sin: NULL
  in_sum: NULL
  in_key: NULL
  in_well: NULL
  out_dir: NULL
  n_threads: NULL
---

<a id="contents"></a>

## Contents

#### [Data Processing](#data_processing)
- [Session Preparation](#session_preparation)
- [Load Inputs](#load_inputs)
- [Assemble Metadata](#assemble data)
- [Fragment Output](#data_output)

#### [QC Metrics and Plots](#qc_metrics)
- [Well-Based QC](#well_qc)
- [10x Genomics Metrics](#well_metrics)
- [Read/UMI/Gene Stats](#rug_stats)
- [Read/UMI/Gene Histograms](#rug_hists)
- [Read/UMI/Gene Scatters](#rug_scatters)
- [Write QC JSON](#json_out)

#### [Session Info](#session_info)

<a id="data_processing"></a>

## Data Processing

<a id="session_preparation"></a>

### Session Preparation

#### Load libraries:
```{r Load Libraries}
start_time <- Sys.time()

quiet_library <- function(...) {
  suppressPackageStartupMessages(library(...))
}
quiet_library(rhdf5)
quiet_library(ATAComb)
quiet_library(H5weaver)
quiet_library(Matrix)
quiet_library(ggplot2)
quiet_library(cowplot)
quiet_library(jsonlite)
```

Declaring start
```{r Declare start}
stm("Starting scATAC-seq preprocessing")
```

#### Argument parsing
```{r Parse arguments}
if(is.null(params$in_h5)) {
  in_fra <- system.file("testdata/outs/fragments.tsv.gz", package = "ATAComb")
  in_sin <- system.file("testdata/outs/singlecell.csv", package = "ATAComb")
  in_sum <- system.file("testdata/outs/summary.csv", package = "ATAComb")
  in_key <- system.file("testdata/test_SampleSheet.csv", package = "ATAComb")
  in_well <- "B000-AP0C1W1"
  out_dir <- tempdir()
  n_threads <- ceiling(parallel::detectCores()/2)
} else {
  in_fra <- params$in_fra
  in_sin <- params$in_sin
  in_sum <- params$in_sum
  in_key <- params$in_key
  in_well <- params$in_well
  out_dir <- params$out_dir
  n_threads <- params$n_threads
}

in_well <- sub("-[AHR]?P","-AP",in_well)

# Check Well ID format
if(!grepl("^[A-Z][0-9]{3}-AP[0-9]C[0-9]W[0-9]$", in_well)) {
  stm(paste0("ERROR: Input Well ID is not in expected format, e.g. B###-AP#C#W# :", in_h5))
  stop()
}

# Check N Threads
if(Sys.info()[["sysname"]] == "Windows") {
  n_threads <- 1
} else if(is.null(n_threads)) {
  n_threads <- ceiling(parallel::detectCores()/2)
} else {
  n_threads <- min(n_threads,
                   parallel::detectCores())
}

stm(paste0("IN  fragments.tsv.gz : ", in_fra))
stm(paste0("IN  singlecell.csv   : ", in_sin))
stm(paste0("IN  summary.csv      : ", in_sum))
stm(paste0("IN  SampleSheet      : ", in_key))
stm(paste0("IN  Well ID          : ", in_well))
stm(paste0("OUT Directory        : ", out_dir))
stm(paste0("PAR N Threads        : ", n_threads))
```

#### Input Parameters
```{r Print Arguments}
print(c(
  paste0("IN  fragments.tsv.gz : ", in_fra),
  paste0("IN  singlecell.csv   : ", in_sin),
  paste0("IN  summary.csv      : ", in_sum),
  paste0("IN  SampleSheet      : ", in_key),
  paste0("IN  Well ID          : ", in_well),
  paste0("OUT H5 directory     : ", out_dir),
  paste0("PAR N Threads        : ", n_threads)
))
```

#### Check Input Files
```{r Check Inputs}
if(!file.exists(in_fra)) {
  stm(paste0("ERROR: Cannot find IN fragments.tsv.gz:", in_fra))
  stop()
}
if(!file.exists(in_sin)) {
  stm(paste0("ERROR: Cannot find IN singlecell.csv:", in_sin))
  stop()
}
if(!file.exists(in_sum)) {
  stm(paste0("ERROR: Cannot find IN summary.csv:", in_sum))
  stop()
}
if(!file.exists(in_key)) {
  stm(paste0("ERROR: Cannot find IN SampleSheet file:", in_key))
  stop()
}
```

#### Create out directory if missing
```{r Create Out Dir}
if(!dir.exists(out_dir)) {
  stm(paste0("Creating Output Directory: ",out_dir))
  dir.create(out_dir, 
             recursive = TRUE)
}
```

#### Declare QC Cutoffs
```{r}
fralt_cut <- 0.7
fritss_cut <- 0.2
frip_cut <- 0.2

stm("Using QC Cutoffs:")
stm(paste("QC FRAlt  : >",fralt_cut))
stm(paste("QC FRITSS : >",fritss_cut))
stm(paste("QC FRIP   : >",frip_cut))
```

```{r}
print(c(
  "Using QC Cutoffs:",
  paste("QC FRAlt  : >",fralt_cut),
  paste("QC FRITSS : >",fritss_cut),
  paste("QC FRIP   : >",frip_cut)
))
```

[Return to Contents](#contents)

<a id="load_inputs"></a>

### Load inputs

#### Load Reference Datasets
```{r Load references}
# Altius Index
alt_idx_file <- system.file("reference/hg38_alt_dhs_index_gr.rds", package = "ATAComb")
stm(paste("Reading Altius Index from", alt_idx_file))
alt_idx_gr <- readRDS(alt_idx_file)

# Buenrostro Peak Reference
buenro_file <- system.file("reference/hg38_GSE123577_filtered_gr.rds", package = "ATAComb")
stm(paste("Reading Buenrostro Peak Reference from", buenro_file))
buenro_gr <- readRDS(buenro_file)

# Gene Body Reference
gene_bodies_file <- system.file("reference/hg38_ensemble93_gene_bodies_gr.rds", package = "ATAComb")
stm(paste("Reading ENSEMBL Gene Body Reference from", gene_bodies_file))
gene_bodies_gr <- readRDS(gene_bodies_file)

# TSS Reference
tss_file <- system.file("reference/hg38_ensemble93_tss_2kb_gr.rds",package = "ATAComb")
stm(paste("Reading ENSEMBL TSS Reference from", tss_file))
tss_2kb_gr <- readRDS(tss_file)
```

#### Load SampleSheet
```{r Load SampleSheet}
stm(paste0("Reading SampleSheet from ", in_key))

samplesheet <- read.csv(in_key)
```

#### Load Fragments.tsv
```{r Load Query}
stm(paste0("Reading fragments from ", in_fra))

fragment_bed_list <- read_10x_fragments(fragments_tsv = in_fra,
                                        singlecell_csv = in_sin,
                                        min_reads = 1e3,
                                        remove_chrM = TRUE,
                                        verbose = FALSE)
```

#### Convert to GRanges
```{r Convert GRanges}
stm(paste0("Converting fragment positions to GenomicRanges"))

fragment_gr_list <- convert_fragments_gr(fragments = fragment_bed_list,
                                         n_threads = n_threads)
```


[Return to Contents](#contents)

<a id="assemble_data"></a>

### Generate Metadata

#### Generate cell IDs
```{r Generate IDs}
stm("Generating Cell IDs")

meta <- data.frame(
  barcodes = ids::uuid(length(fragment_gr_list),
                       drop_hyphens = TRUE, 
                       use_time = TRUE),
  cell_name = ids::adjective_animal(length(fragment_gr_list),
                                    n_adjectives = 2,
                                    max_len = 10),
  original_barcodes = sub("-1","",names(fragment_gr_list)),
  stringsAsFactors = FALSE)

names(fragment_gr_list) <- meta$barcodes
```

#### Add Well/Chip/Pool/Sample IDs
```{r}
stm("Adding Batch and Sample ID metadata")
meta$well_id <- in_well
meta$chip_id <- sub("W[0-9]+$","",meta$well_id)
meta$batch_id <- sub("-.+","",meta$chip_id)
meta$pbmc_sample_id <- as.character(samplesheet$SampleID[samplesheet$WellID == in_well])
```

#### Compute QC Stats
```{r}
meta$n_reads <- unlist(lapply(fragment_gr_list, length))

ritss <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                            target_GRanges = tss_2kb_gr,
                            binarize = TRUE,
                            sparse = FALSE,
                            aggregate = TRUE,
                            n_threads = n_threads)

meta$fritss <- ritss / meta$n_reads

nontss_peak_gr <- filter_gr(buenro_gr,
                            tss_2kb_gr)

rip <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                         target_GRanges = nontss_peak_gr,
                         binarize = TRUE,
                         sparse = FALSE,
                         aggregate = TRUE,
                         n_threads = n_threads)

meta$frip <- rip / meta$n_reads

reduced_alt <- reduce(alt_idx_gr)
ralt <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                          target_GRanges = reduced_alt,
                          binarize = TRUE,
                          sparse = FALSE,
                          aggregate = TRUE,
                          n_threads = n_threads)

meta$fralt <- ralt / meta$n_reads
```


#### Filter based on cutoffs
```{r}
stm("Filtering barcodes based on QC Cutoffs")
filtered_meta <- meta
filtered_meta <- filtered_meta[filtered_meta$fralt  > fralt_cut ,]
filtered_meta <- filtered_meta[filtered_meta$fritss > fritss_cut,]
filtered_meta <- filtered_meta[filtered_meta$frip   > frip_cut  ,]

fragment_gr_list <- fragment_gr_list[filtered_meta$barcodes]
```

#### Output filtered fragments.tsv.gz
```{r}
stm("Building filtered_fragments.tsv.gz")
filtered_bed_list <- convert_fragments_bed(fragment_gr_list,
                                           n_threads = n_threads)
n_fragments <- unlist(lapply(filtered_bed_list, nrow))

filtered_bed <- do.call("rbind",filtered_bed_list)
filtered_bed$barcode <- rep(names(filtered_bed_list), n_fragments)
rm(filtered_bed_list)

out_frag_file <- file.path(out_dir, paste0(in_well, "_filtered_fragments.tsv.gz"))
stm(paste0("Writing filtered fragments to: ", out_frag_file))
fwrite(filtered_bed,
       file = out_frag_file)

rm(filtered_bed)
```

[Return to Contents](#contents)

<a id="mats"></a>

## Compute Feature Matrices

<a id="mats_alt"></a>

#### Compute overlaps with Altius index
```{r Altius overlaps}
stm("Computing overlaps with Altius Index Features")

alt_ol_mat <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                                target_GRanges = alt_idx_gr,
                                binarize = FALSE,
                                sparse = TRUE,
                                aggregate = FALSE,
                                n_threads = n_threads)

rownames(alt_ol_mat) <- paste(seqnames(alt_idx_gr),
                              start(alt_idx_gr),
                              end(alt_idx_gr),
                              sep = "_")
```

#### Build Altius list and output .h5 file
```{r Build Altius h5}
alt_list <- list(matrix_dgCMatrix = alt_ol_mat)
rm(alt_ol_mat)

alt_list <- h5_list_convert_from_dgCMatrix(alt_list,
                                           target = "matrix")
alt_list$matrix$observations <- as.list(filtered_meta)
```

Write Altius Index .h5 file
```{r Write Altius h5}
alt_out_file <- file.path(out_dir, paste0(in_well, "_altius.h5"))
stm(paste("Writing Altius Index h5 file to", alt_out_file))
write_h5_list(alt_list,
              alt_out_file)

rm(alt_list)
```

[Return to Contents](#contents)

<a id="mats_buenro"></a>

#### Compute overlaps with Buenrostro reference
```{r Buenrostro overlaps}
stm("Computing overlaps with Buenrostro reference Features")

buenro_ol_mat <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                                   target_GRanges = buenro_gr,
                                   binarize = FALSE,
                                   sparse = TRUE,
                                   aggregate = FALSE,
                                   n_threads = n_threads)

rownames(buenro_ol_mat) <- paste(seqnames(buenro_gr),
                                 start(buenro_gr),
                                 end(buenro_gr),
                                 sep = "_")

```

#### Build Buenrostro list and output .h5 file
```{r Build Buenrostro h5}
buenro_list <- list(matrix_dgCMatrix = buenro_ol_mat)
rm(buenro_ol_mat)

buenro_list <- h5_list_convert_from_dgCMatrix(buenro_list,
                                           target = "matrix")
buenro_list$matrix$observations <- as.list(filtered_meta)
```

Write Buenrostro .h5 file
```{r Write Buenrostro h5}
buenro_out_file <- file.path(out_dir, paste0(in_well, "_buenrostro_peak.h5"))
stm(paste("Writing Buenrostro Peak h5 file to", buenro_out_file))
write_h5_list(buenro_list,
              buenro_out_file)

rm(buenro_list)
```

[Return to Contents](#contents)

<a id="mats_gene_body"></a>

#### Gene body overlaps
```{r Gene Overlaps}
stm("Computing overlaps with ENSEMBL reference Features")

gene_body_ol_mat <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                                                       target_GRanges = gene_bodies_gr,
                                                       binarize = FALSE,
                                                       sparse = TRUE,
                                                       aggregate = FALSE,
                                                       n_threads = n_threads)
rownames(gene_body_ol_mat) <- gene_bodies_gr$gene_id
```

#### Build Gene Body list and output .h5 file
```{r Build Gene Body h5}
gene_body_list <- list(matrix_dgCMatrix = gene_body_ol_mat)
rm(gene_body_ol_mat)

gene_body_list <- h5_list_convert_from_dgCMatrix(gene_body_list,
                                           target = "matrix")
gene_body_list$matrix$observations <- as.list(filtered_meta)
```

Write Gene Body .h5 file
```{r Write Gene Body h5}
gene_body_out_file <- file.path(out_dir, paste0(in_well, "_gene_body.h5"))
stm(paste("Writing Gene Body h5 file to", gene_body_out_file))
write_h5_list(gene_body_list,
              gene_body_out_file)

rm(gene_body_list)
```

[Return to Contents](#contents)

<a id="mats_tss"></a>

#### TSS Overlaps
```{r TSS Overlaps}
tss_2kb_ol_mat <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                                                       target_GRanges = tss_2kb_gr,
                                                       binarize = FALSE,
                                                       sparse = TRUE,
                                                       aggregate = FALSE,
                                                       n_threads = n_threads)
rownames(tss_2kb_ol_mat) <- tss_2kb_gr$gene_id
```

#### Build TSS list and output .h5 file
```{r Build TSS h5}
tss_2kb_list <- list(matrix_dgCMatrix = tss_2kb_ol_mat)
rm(tss_2kb_ol_mat)

tss_2kb_list <- h5_list_convert_from_dgCMatrix(tss_2kb_list,
                                           target = "matrix")
tss_2kb_list$matrix$observations <- as.list(filtered_meta)
```

Write Gene Body .h5 file
```{r Write TSS h5}
tss_2kb_out_file <- file.path(out_dir, paste0(in_well, "_tss_2kb.h5"))
stm(paste("Writing TSS h5 file to", tss_2kb_out_file))
write_h5_list(tss_2kb_list,
              tss_2kb_out_file)

rm(tss_2kb_list)
```

[Return to Contents](#contents)

<a id="mats_5kb"></a>

#### Compute 5kb window matrix
```{r 5kb Window Matrices}
stm("Converting fragment positions to windows")

window_5k_mat <- convert_fragments_windows(fragments = fragment_bed_list,
                                           window_size = 5e3,
                                           genome = "hg38",
                                           n_threads = n_threads)
```

#### Build TSS list and output .h5 file
```{r Build 5k Window h5}
window_5k_list <- list(matrix_dgCMatrix = window_5k_mat)
rm(window_5k_mat)

window_5k_list <- h5_list_convert_from_dgCMatrix(window_5k_list,
                                           target = "matrix")
window_5k_list$matrix$observations <- as.list(filtered_meta)
```

Write Gene Body .h5 file
```{r Write 5k Window h5}
window_5k_out_file <- file.path(out_dir, paste0(in_well, "_window_5kb.h5"))
stm(paste("Writing 5kb Window h5 file to", window_5k_out_file))
write_h5_list(window_5k_list,
              window_5k_out_file)

rm(window_5k_list)
```

[Return to Contents](#contents)

<a id="mats_20kb"></a>

```{r 20kb Window Matrices}
window_20k_mat <- convert_fragments_windows(fragments = fragment_bed_list,
                                            window_size = 2e4,
                                            genome = "hg38",
                                            n_threads = n_threads)

```

#### Build TSS list and output .h5 file
```{r Build 20k Window h5}
window_20k_list <- list(matrix_dgCMatrix = window_20k_mat)
rm(window_20k_mat)

window_20k_list <- h5_list_convert_from_dgCMatrix(window_20k_list,
                                           target = "matrix")
window_20k_list$matrix$observations <- as.list(filtered_meta)
```

Write Gene Body .h5 file
```{r Write 20k Window h5}
window_20k_out_file <- file.path(out_dir, paste0(in_well, "_window_20kb.h5"))
stm(paste("Writing 20kb Window h5 file to", window_20k_out_file))
write_h5_list(window_20k_list,
              window_20k_out_file)

rm(window_20k_list)
```

[Return to Contents](#contents)

<a id="qc_plots"></a>

#### Plot QC Stats
```{r}
n_reads_hist <- qc_hist_plot(meta,
                             column = "n_reads",
                             name_x = "N Reads per Cell",
                             log_x = TRUE,
                             fill = "dodgerblue",
                             target = 2500)
fralt_hist <- qc_frac_hist_plot(meta,
                                column = "fralt",
                                name_x = "Frac Reads in Altius Index (FRAlt)",
                                fill = "darkgreen",
                                target = fralt_cut)
fritss_hist <- qc_frac_hist_plot(meta,
                                 column = "fritss",
                                 name_x = "Frac Reads in TSS (FRITSS)",
                                 fill = "orangered",
                                 target = fritss_cut)
frip_hist <- qc_frac_hist_plot(meta,
                               column = "frip",
                               name_x = "Frac Reads in Peaks (FRIP)",
                               fill = "mediumorchid3",
                               target = frip_cut)


plot_list <- list(n_reads_hist,
                  fralt_hist,
                  fritss_hist,
                  frip_hist)

cowplot::plot_grid(plotlist = plot_list,
                   nrow = 4, ncol = 1)
```

#### Pairwise scatters
```{r}
qc_scatter_plot(meta,
                column_x = "n_reads",
                name_x = "N Reads per Cell",
                column_y = "fralt",
                name_y = "Frac Reads in Altius Index (FRAlt)",
                log_x = TRUE, log_y = FALSE, frac_y = TRUE,
                show_targets = FALSE) +
  geom_vline(aes(xintercept = 2.5e3), linetype = "dashed", size = 0.2) +
  geom_hline(aes(yintercept = fralt_cut), linetype = "dashed", size = 0.2)
```

```{r}
qc_scatter_plot(meta,
                column_x = "fritss",
                name_x = "Frac Reads in TSS (FRITSS)",
                column_y = "frip",
                name_y = "Frac Reads in Peaks (FRIP)",
                log_x = FALSE, frac_x = TRUE, log_y = FALSE, frac_y = TRUE,
                show_targets = FALSE,
                color = "orange") +
  geom_vline(aes(xintercept = fritss_cut), linetype = "dashed", size = 0.2) +
  geom_hline(aes(yintercept = frip_cut), linetype = "dashed", size = 0.2)
```



#### Filtered QC Hists
```{r}
n_reads_hist <- qc_hist_plot(filtered_meta,
                             column = "n_reads",
                             name_x = "N Reads per Cell",
                             log_x = TRUE,
                             fill = "dodgerblue",
                             target = 2500)
fralt_hist <- qc_frac_hist_plot(filtered_meta,
                                column = "fralt",
                                name_x = "Frac Reads in Altius Index (FRAlt)",
                                fill = "darkgreen",
                                target = fralt_cut)
fritss_hist <- qc_frac_hist_plot(filtered_meta,
                                 column = "fritss",
                                 name_x = "Frac Reads in TSS (FRITSS)",
                                 fill = "orangered",
                                 target = fritss_cut)
frip_hist <- qc_frac_hist_plot(filtered_meta,
                               column = "frip",
                               name_x = "Frac Reads in Peaks (FRIP)",
                               fill = "mediumorchid3",
                               target = frip_cut)


plot_list <- list(n_reads_hist,
                  fralt_hist,
                  fritss_hist,
                  frip_hist)

cowplot::plot_grid(plotlist = plot_list,
                   nrow = 4, ncol = 1)
```

[Return to Contents](#contents)

<a id="json_out"></a>

### Write QC JSON

```{r Save QC JSON}
stm("No JSON available at this point.")
```

[Return to Contents](#contents)

<a id="session_info"></a>

## Session Information

```{r Session Info}
sessionInfo()
```

Total time elapsed
```{r Show Time}
end_time <- Sys.time()
diff_time <- end_time - start_time
time_message <- paste0("Elapsed Time: ", 
                       round(diff_time, 3),
                       " ", units(diff_time))
print(time_message)
stm(time_message)
stm("H5 metadata process complete.")
```

[Return to Contents](#contents)
