---
title: "Preprocess 10x Genomics scATAC-seq data"
author: 
 - Lucas Graybuck
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    code_folding: hide
    df_print: paged
    self_contained: true
params:
  in_fra: NULL
  in_sin: NULL
  in_sum: NULL
  in_key: NULL
  in_well: NULL
  out_dir: NULL
  n_threads: NULL
---

<a id="contents"></a>

## Contents

#### [Data Processing](#data_processing)
- [Session Preparation](#session_preparation)
- [Load Inputs](#load_inputs)
- [Assemble Metadata](#assemble data)
- [Data Output](#data_output)

#### [QC Metrics and Plots](#qc_metrics)
- [Well-Based QC](#well_qc)
- [10x Genomics Metrics](#well_metrics)
- [Read/UMI/Gene Stats](#rug_stats)
- [Read/UMI/Gene Histograms](#rug_hists)
- [Read/UMI/Gene Scatters](#rug_scatters)
- [Write QC JSON](#json_out)

#### [Session Info](#session_info)

<a id="data_processing"></a>

## Data Processing

<a id="session_preparation"></a>

### Session Preparation

#### Load libraries:
```{r Load Libraries}
start_time <- Sys.time()

quiet_library <- function(...) {
  suppressPackageStartupMessages(library(...))
}
quiet_library(rhdf5)
quiet_library(ATAComb)
quiet_library(H5weaver)
quiet_library(Matrix)
quiet_library(ggplot2)
quiet_library(cowplot)
quiet_library(jsonlite)
```

Declaring start
```{r Declare start}
stm("Starting scATAC-seq preprocessing")
```

#### Argument parsing
```{r Parse arguments}
if(is.null(params$in_h5)) {
  in_fra <- system.file("testdata/outs/fragments.tsv.gz", package = "ATAComb")
  in_sin <- system.file("testdata/outs/singlecell.csv", package = "ATAComb")
  in_sum <- system.file("testdata/outs/summary.csv", package = "ATAComb")
  in_key <- system.file("reference/SampleSheet_fallback.csv", package = "H5weaver")
  in_well <- "B000-P0C0W0"
  out_dir <- tempdir()
  n_threads <- ceiling(parallel::detectCores()/2)
} else {
  in_fra <- params$in_fra
  in_sin <- params$in_sin
  in_sum <- params$in_sum
  in_key <- params$in_key
  in_well <- params$in_well
  out_dir <- params$out_dir
  n_threads <- params$n_threads
}

in_well <- sub("-[AHR]?P","-AP",in_well)

# Check Well ID format
if(!grepl("^[A-Z][0-9]{3}-AP[0-9]C[0-9]W[0-9]$", in_well)) {
  stm(paste0("ERROR: Input Well ID is not in expected format, e.g. B###-AP#C#W# :", in_h5))
  stop()
}

# Check N Threads
if(Sys.info()[["sysname"]] == "Windows") {
  n_threads <- 1
} else if(is.null(n_threads)) {
  n_threads <- ceiling(parallel::detectCores()/2)
} else {
  n_threads <- min(n_threads,
                   parallel::detectCores())
}

stm(paste0("IN  fragments.tsv.gz : ", in_fra))
stm(paste0("IN  singlecell.csv   : ", in_sin))
stm(paste0("IN  summary.csv      : ", in_sum))
stm(paste0("IN  SampleSheet      : ", in_key))
stm(paste0("IN  Well ID          : ", in_well))
stm(paste0("OUT Directory        : ", out_dir))
stm(paste0("PAR N Threads        : ", n_threads))
```

#### Input Parameters
```{r Print Arguments}
print(c(
  paste0("IN  fragments.tsv.gz : ", in_fra),
  paste0("IN  singlecell.csv   : ", in_sin),
  paste0("IN  summary.csv      : ", in_sum),
  paste0("IN  SampleSheet      : ", in_key),
  paste0("IN  Well ID          : ", in_well),
  paste0("OUT H5 directory     : ", out_dir),
  paste0("PAR N Threads        : ", n_threads)
))
```

#### Check Input Files
```{r Check Inputs}
if(!file.exists(in_fra)) {
  stm(paste0("ERROR: Cannot find IN fragments.tsv.gz:", in_fra))
  stop()
}
if(!file.exists(in_sin)) {
  stm(paste0("ERROR: Cannot find IN singlecell.csv:", in_sin))
  stop()
}
if(!file.exists(in_sum)) {
  stm(paste0("ERROR: Cannot find IN summary.csv:", in_sum))
  stop()
}
if(!file.exists(in_key)) {
  stm(paste0("ERROR: Cannot find IN SampleSheet file:", in_key))
  stop()
}
```

#### Create out directory if missing
```{r Create Out Dir}
if(!dir.exists(out_dir)) {
  stm(paste0("Creating Output Directory: ",out_dir))
  dir.create(out_dir, 
             recursive = TRUE)
}
```


[Return to Contents](#contents)

<a id="load_inputs"></a>

### Load inputs

#### Load Fragments.tsv
```{r Load Query}
stm(paste0("Reading fragments from ", in_fra))

fragment_bed_list <- read_10x_fragments(fragments_tsv = in_fra,
                                        singlecell_csv = in_sin,
                                        min_reads = 1e3,
                                        remove_chrM = TRUE,
                                        verbose = FALSE)
```

#### Convert to GRanges
```{r Convert GRanges}
stm(paste0("Converting fragment positions to GenomicRanges"))

fragment_gr_list <- convert_fragments_gr(fragments = fragment_bed_list,
                                         n_threads = n_threads)
```

#### Compute overlaps with Altius index
```{r Altius overlaps}
stm("Reading Altius Index from ATAComb")

alt_idx_gr <- readRDS(system.file("reference/hg38_alt_dhs_index_gr.rds", package = "ATAComb"))

stm("Computing overlaps with Altius Index")

alt_ol_mat <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                                target_GRanges = alt_idx_gr,
                                binarize = FALSE,
                                sparse = TRUE,
                                aggregate = FALSE,
                                n_threads = n_threads)

rownames(alt_ol_mat) <- paste(seqnames(alt_idx_gr),
                              start(alt_idx_gr),
                              end(alt_idx_gr),
                              sep = "_")
```

#### Compute overlaps with Buenrostro reference
```{r Buenrostro overlaps}
stm("Computing overlaps with Buenrostro reference")

buenro_gr <- readRDS(system.file("reference/hg38_GSE123577_filtered_gr.rds", package = "ATAComb"))

buenro_ol_mat <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                                   target_GRanges = buenro_gr,
                                   binarize = FALSE,
                                   sparse = TRUE,
                                   aggregate = FALSE,
                                   n_threads = n_threads)

rownames(buenro_ol_mat) <- paste(seqnames(buenro_gr),
                                 start(buenro_gr),
                                 end(buenro_gr),
                                 sep = "_")

```

#### Gene body and TSS overlaps
```{r Gene Overlaps}
stm("Computing overlaps with ENSEMBL reference")
gene_bodies_gr <- readRDS(system.file("reference/hg38_ensemble93_gene_bodies_gr.rds",
                                      package = "ATAComb"))

gene_body_ol_mat <- buenro_ol_mat <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                                                       target_GRanges = gene_bodies_gr,
                                                       binarize = FALSE,
                                                       sparse = TRUE,
                                                       aggregate = FALSE,
                                                       n_threads = n_threads)
rownames(gene_body_ol_mat) <- gene_bodies_gr$gene_id

tss_2kb_gr <- readRDS(system.file("reference/hg38_ensemble93_tss_2kb_gr.rds",
                                   package = "ATAComb"))

tss_2kb_ol_mat <- buenro_ol_mat <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                                                       target_GRanges = tss_2kb_gr,
                                                       binarize = FALSE,
                                                       sparse = TRUE,
                                                       aggregate = FALSE,
                                                       n_threads = n_threads)
rownames(tss_2kb_ol_mat) <- tss_2kb_gr$gene_id


```


```{r}
n_reads <- unlist(lapply(fragment_gr_list, length))

ritss <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                            target_GRanges = tss_2kb_gr,
                            binarize = TRUE,
                            sparse = FALSE,
                            aggregate = TRUE,
                            n_threads = n_threads)

fritss <- ritss / n_reads

nontss_peak_gr <- filter_gr(buenro_gr,
                            tss_2kb_gr)

rip <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                         target_GRanges = nontss_peak_gr,
                         binarize = TRUE,
                         sparse = FALSE,
                         aggregate = TRUE,
                         n_threads = n_threads)

frip <- rip / n_reads

reduced_alt <- reduce(alt_idx_gr)
ralt <- count_frag_ol_ref(query_fragments = fragment_gr_list,
                          target_GRanges = reduced_alt,
                          binarize = TRUE,
                          sparse = FALSE,
                          aggregate = TRUE,
                          n_threads = n_threads)

fralt <- ralt / n_reads
```


#### Compute Window-based matrix
```{r Window Matrices}
stm("Converting fragment positions to windows")

window_5k_mat <- convert_fragments_windows(fragments = fragment_bed_list,
                                           window_size = 5e3,
                                           genome = "hg38",
                                           n_threads = n_threads)

window_20k_mat <- convert_fragments_windows(fragments = fragment_bed_list,
                                            window_size = 2e4,
                                            genome = "hg38",
                                            n_threads = n_threads)

```

```{r}
window_cov_5k <- colSums(window_5k_mat > 0) / nrow(window_5k_mat)
window_cov_20k <- colSums(window_20k_mat > 0) / nrow(window_20k_mat)
```


#### Assemble h5 list
```{r Assemble h5 list}
h5_list <- list(alt_index_dgCMatrix = alt_ol_mat,
                buenro_peaks_dgCMatrix = buenro_ol_mat,
                windows_5k_dgCMatrix = window_5k_mat,
                windows_20k_dgCMatrix = window_20k_mat)

h5_list <- h5_list_convert_from_dgCMatrix(h5_list, target = "alt_index")
h5_list <- h5_list_convert_from_dgCMatrix(h5_list, target = "buenro_peaks")
h5_list <- h5_list_convert_from_dgCMatrix(h5_list, target = "windows_5k")
h5_list <- h5_list_convert_from_dgCMatrix(h5_list, target = "windows_20k")
```

#### Add cell ids
```{r Add cell ids}
h5_list <- add_cell_ids(h5_list, 
                        target = "windows_5k",
                        add_uuid = TRUE,
                        replace_barcode = TRUE,
                        retain_original_barcode = TRUE,
                        add_name = TRUE)

h5_list$alt_index$barcodes <- h5_list$windows_5k$barcodes
h5_list$buenro_peaks$barcodes <- h5_list$windows_5k$barcodes
h5_list$windows_20k$barcodes <- h5_list$windows_5k$barcodes
```

#### Count TSS overlaps


[Return to Contents](#contents)

<a id="assemble_data"></a>

### Assemble data

#### Compute N UMIs and N Genes per cell
```{r n_umi and n_genes}
stm("Computing UMI and Gene Counts per Cell")

h5_list <- h5_list_convert_to_dgCMatrix(h5_list, target = "matrix")

h5_list <- set_list_path(h5_list,
                         "/matrix/observations/n_umis",
                         unname(colSums(h5_list$matrix_dgCMatrix)))

h5_list <- set_list_path(h5_list,
                         "/matrix/observations/n_genes",
                         unname(colSums(h5_list$matrix_dgCMatrix > 0)))

h5_list <- h5_list_convert_from_dgCMatrix(h5_list, target = "matrix")
```

#### Add cell ids
```{r}
stm("Adding Cell UUIDs and Names")

h5_list <-add_cell_ids(h5_list,
                       add_uuid = TRUE,
                       replace_barcode = TRUE,
                       retain_original_barcode = TRUE,
                       add_name = TRUE)
```

#### Generate Chip, Pool, and Batch IDs based on well_id
```{r Chip Pool Batch}
stm("Adding Batch, Pool, Chip, and Well metadata")

h5_list <- add_well_metadata(h5_list,
                             well_id = in_well)
```

#### Add chrM gene counts
```{r chrM UMIs}
stm("Adding chrM count metadata")

h5_list <- h5_list_add_mito_umis(h5_list)
```

#### Add Well Metrics
```{r Add Well Metrics}
well_metrics <- read_tenx_metrics(in_sum)

well_metrics <- as.list(well_metrics)

h5_list <- set_list_path(h5_list,
                         "/well",
                         well_metrics)

h5_list <- set_list_path(h5_list,
                         "/well/well_id",
                         in_well)
```


[Return to Contents](#contents)

<a id="data_output"></a>

### Write Output

#### Write HDF5 files
```{r Write split files}
stm(paste0("Writing HDF5 to ", out_h5))
write_h5_list(h5_list,
              h5_file = out_h5,
              overwrite = TRUE)
h5closeAll()
```

[Return to Contents](#contents)

<a id="qc_metrics"></a>

## QC Tables and Plots

Extract metadata for plotting
```{r Assemble Metadata}
stm("Generating tables and plots for report")
meta <- h5_list_cell_metadata(h5_list)

qc_list <- list(report_type = "add_tenx_rna_metadata",
                report_datetime = as.character(start_time),
                report_uuid = ids::uuid(use_time = TRUE),
                package = "H5weaver",
                package_version = sessionInfo()$otherPkgs$H5weaver$Version,
                well_id = in_well)
```

[Return to Contents](#contents)

<a id="well_qc"></a>

### Well QC

<a id="well_metrics"></a>

#### 10x Genomics Mapping Metrics
```{r Well Metrics}
qc_well_metrics <- data.frame(tenx_metric = names(well_metrics),
                              value = as.vector(unlist(well_metrics)),
                              stringsAsFactors = FALSE)

qc_well_metrics$value[qc_well_metrics$tenx_metric == "number_of_reads"] <- as.numeric(qc_well_metrics$value[qc_well_metrics$tenx_metric == "number_of_reads"] / 1e6)
qc_well_metrics$tenx_metric[qc_well_metrics$tenx_metric == "number_of_reads"] <- "millions_of_reads"

qc_list$well_metrics <- split(qc_well_metrics$value, 1:nrow(qc_well_metrics))
names(qc_list$well_metrics) <- qc_well_metrics$tenx_metric

qc_table(qc_well_metrics)
```

[Return to Contents](#contents)

<a id="rug_stats"></a>

#### Well Read/UMI/Gene Summary Stats
```{r Read UMI Gene Stats}
overview_stats <- data.frame(statistic = c("N Mapped Reads",
                                           "N UMIs",
                                           "N Mito. UMIs",
                                           "N Genes",
                                           "Reads per UMI",
                                           "UMIs per Gene",
                                           "Reads per Gene"),
                             q_25 = c(quantile(meta$n_reads, 0.25),
                                      quantile(meta$n_umis,  0.25),
                                      quantile(meta$n_mito_umis, 0.25),
                                      quantile(meta$n_genes, 0.25),
                                      quantile(meta$n_reads/meta$n_umis, 0.25, na.rm = TRUE),
                                      quantile(meta$n_umis/meta$n_genes, 0.25, na.rm = TRUE),
                                      quantile(meta$n_reads/meta$n_genes, 0.25, na.rm = TRUE)),
                             median = c(quantile(meta$n_reads, 0.5),
                                        quantile(meta$n_umis,  0.50),
                                        quantile(meta$n_mito_umis, 0.50),
                                        quantile(meta$n_genes, 0.50),
                                        quantile(meta$n_reads/meta$n_umis, 0.50, na.rm = TRUE),
                                        quantile(meta$n_umis/meta$n_genes, 0.50, na.rm = TRUE),
                                        quantile(meta$n_reads/meta$n_genes, 0.50, na.rm = TRUE)),
                             q_75 = c(quantile(meta$n_reads, 0.75),
                                      quantile(meta$n_umis,  0.75),
                                      quantile(meta$n_mito_umis, 0.25),
                                      quantile(meta$n_genes, 0.75),
                                      quantile(meta$n_reads/meta$n_umis, 0.75, na.rm = TRUE),
                                      quantile(meta$n_umis/meta$n_genes, 0.75, na.rm = TRUE),
                                      quantile(meta$n_reads/meta$n_genes, 0.75, na.rm = TRUE)))

overview_stats$q_25 <- round(overview_stats$q_25, 2)
overview_stats$median <- round(overview_stats$median, 2)
overview_stats$q_75 <- round(overview_stats$q_75, 2)

qc_list$well_count_stats <- lapply(1:nrow(overview_stats),
                                   function(x) {
                                     list(q_25 = overview_stats$q_25[x],
                                          median = overview_stats$median[x],
                                          q_75 = overview_stats$q_75[x])
                                   })
names(qc_list$well_count_stats) <- tolower(gsub("[\\. ]+", "_", overview_stats$statistic))

qc_table(overview_stats)
```

[Return to Contents](#contents)

<a id="rug_hists"></a>

### Well Read/UMI/Gene Histograms
```{r QC histograms}
reads_plot <- qc_hist_plot(meta,
                           column = "n_reads",
                           name_x = "N Reads per Cell",
                           fill = "dodgerblue",
                           target = 2e4) +
  ggtitle("Alignment/Mapping Distributions")

umis_plot <- qc_hist_plot(meta,
                          column = "n_umis",
                          name_x = "N UMIs per Cell",
                          fill = "purple",
                          target = 8e3)

genes_plot <- qc_hist_plot(meta,
                           column = "n_genes",
                           name_x = "N Genes per Cell",
                           fill = "orangered",
                           target = 2e3)

histogram_list <- list(reads_plot,
                       umis_plot,
                       genes_plot)

suppressWarnings(
  plot_grid(plotlist = histogram_list,
            ncol = 1,
            nrow = 3)
)
```

[Return to Contents](#contents)

<a id="rug_scatters"></a>

#### UMIs vs Reads Scatter Plot
```{r Read UMI Gene Scatter Plots}
umi_reads_scatter <- qc_scatter_plot(meta,
                                     column_x = "n_reads",
                                     name_x = "N Reads per Cell",
                                     column_y = "n_umis",
                                     name_y = "N UMIs per Cell",
                                     color = "darkblue") +
  ggtitle("UMIs vs Reads")

suppressWarnings(
  umi_reads_scatter
)
```

#### Genes vs UMIs Scatter Plot
```{r}
genes_umis_scatter <- qc_scatter_plot(meta,
                                      column_x = "n_umis",
                                      name_x = "N UMIs per Cell",
                                      column_y = "n_genes",
                                      name_y = "N Genes per Cell",
                                      color = "red") +
  ggtitle("Genes vs UMIs")

suppressWarnings(
  genes_umis_scatter
)
```

#### Genes vs Reads Scatter Plot
```{r}
genes_reads_scatter <- qc_scatter_plot(meta,
                                       column_x = "n_reads",
                                       name_x = "N Reads per Cell",
                                       column_y = "n_genes",
                                       name_y = "N Genes per Cell",
                                       color = "darkgreen") +
  ggtitle("Genes vs Reads")

suppressWarnings(
  genes_reads_scatter
)
```

[Return to Contents](#contents)

<a id="json_out"></a>

### Write QC JSON

```{r Save QC JSON}
stm(paste0("Writing JSON to ", out_json))

qc_list_json <- jsonlite::toJSON(qc_list,
                                 auto_unbox = TRUE,
                                 pretty = TRUE)

writeLines(qc_list_json,
           out_json)

```

[Return to Contents](#contents)

<a id="session_info"></a>

## Session Information

```{r Session Info}
sessionInfo()
```

Total time elapsed
```{r Show Time}
end_time <- Sys.time()
diff_time <- end_time - start_time
time_message <- paste0("Elapsed Time: ", 
                       round(diff_time, 3),
                       " ", units(diff_time))
print(time_message)
stm(time_message)
stm("H5 metadata process complete.")
```

[Return to Contents](#contents)
