---
title: "10x Genomics scATAC-seq QC"
author: 
 - Lucas Graybuck
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    code_folding: hide
    df_print: paged
    self_contained: true
params:
  in_pre: NULL
  in_sin: NULL
  in_sum: NULL
  in_key: NULL
  in_well: NULL
  out_dir: NULL
---

<a id="contents"></a>

## Contents

#### [Data Processing](#data_processing)
- [Session Preparation](#session_preparation)
- [Load References](#load_refs)
- [Assemble Metadata](#assemble_meta)
- [Generate Matrices](#mats)

#### [QC Metrics](#qc_stats)
- [Cell Barcode QC](#barcode_stats)
- [Fragment Metrics](#fragment_stats)
- [Saturation Metrics](#saturation_stats)
- [10x Genomics Metrics](#tenx_stats)

#### [QC Plots](#qc_plots)
- [UMI Saturation Plot](#saturation_plot)
- [Signal Saturation Plot](#signal_plot)
- [Fragment Width Plot](#width_plot)
- [All Cell Histograms](#all_hist_plot)
- [Reads vs FRAlt Plot](#reads_fralt_plot)
- [Reads vs FRIP Plot](#reads_frip_plot)
- [FRIP vs FRITSS Plot](#frip_fritss_plot)
- [RIP vs RITSS Plot](#rip_ritss_plot)
- [Filtered Histograms](#filtered_hist_plot)

[Write QC JSON](#json_out)

#### [Session Info](#session_info)

<a id="data_processing"></a>

## Data Processing

<a id="session_preparation"></a>

### Session Preparation

#### Load libraries:
```{r Load Libraries}
start_time <- Sys.time()

quiet_library <- function(...) {
  suppressPackageStartupMessages(library(...))
}
quiet_library(rhdf5)
quiet_library(ATAComb)
quiet_library(H5weaver)
quiet_library(Matrix)
quiet_library(S4Vectors)
quiet_library(ggplot2)
quiet_library(cowplot)
quiet_library(jsonlite)
options(stringsAsFactors = FALSE)
```

Declaring start
```{r Declare start}
stm("Starting scATAC-seq preprocessing")
```

#### Argument parsing
```{r Parse arguments}
if(is.null(params$in_pre)) {
  in_pre <- system.file("testdata/preprocessed", package = "ATAComb")
  in_sin <- system.file("testdata/outs/singlecell.csv", package = "ATAComb")
  in_sum <- system.file("testdata/outs/summary.csv", package = "ATAComb")
  in_key <- system.file("testdata/test_SampleSheet.csv", package = "ATAComb")
  in_well <- "B000-AP0C1W1"
  out_dir <- tempdir()
} else {
  in_pre <- params$in_pre
  in_sin <- params$in_sin
  in_sum <- params$in_sum
  in_key <- params$in_key
  in_well <- params$in_well
  out_dir <- params$out_dir
}

in_well <- sub("-[AHR]?P","-AP",in_well)
in_batch <- sub("-.+","",in_well)

# Check Well ID format
if(!grepl("^[A-Z][0-9]{3}-AP[0-9]C[0-9]W[0-9]$", in_well)) {
  stm(paste0("ERROR: Input Well ID is not in expected format, e.g. B###-AP#C#W# :", in_h5))
  stop()
}

stm(paste0("IN  preprocessed dir : ", in_pre))
stm(paste0("IN  singlecell.csv   : ", in_sin))
stm(paste0("IN  summary.csv      : ", in_sum))
stm(paste0("IN  SampleSheet      : ", in_key))
stm(paste0("IN  Well ID          : ", in_well))
stm(paste0("OUT Directory        : ", out_dir))
```

#### Input Parameters
```{r Print Arguments}
print(c(
  paste0("IN  preprocessed dir : ", in_pre),
  paste0("IN  singlecell.csv   : ", in_sin),
  paste0("IN  summary.csv      : ", in_sum),
  paste0("IN  SampleSheet      : ", in_key),
  paste0("IN  Well ID          : ", in_well),
  paste0("OUT H5 directory     : ", out_dir)
))
```

#### Check Input Files
```{r Check Inputs}
if(!dir.exists(in_pre)) {
  stm(paste0("ERROR: Cannot find IN preprocessed dir:", in_pre))
  stop()
}
if(!file.exists(in_sin)) {
  stm(paste0("ERROR: Cannot find IN singlecell.csv:", in_sin))
  stop()
}
if(!file.exists(in_sum)) {
  stm(paste0("ERROR: Cannot find IN summary.csv:", in_sum))
  stop()
}
if(!file.exists(in_key)) {
  stm(paste0("ERROR: Cannot find IN SampleSheet file:", in_key))
  stop()
}
in_prefix <- file.path(in_pre, in_well)
```

#### Create out directory if missing
```{r Create Out Dir}
if(!dir.exists(out_dir)) {
  stm(paste0("Creating Output Directory: ",out_dir))
  dir.create(out_dir, 
             recursive = TRUE)
}
```

#### Declare QC Cutoffs
```{r}
fralt_cut <- 0.5
fritss_cut <- 0.2
frip_cut <- 0.2

stm("Using QC Cutoffs:")
stm(paste("QC FRAlt  : >",fralt_cut))
stm(paste("QC FRITSS : >",fritss_cut))
stm(paste("QC FRIP   : >",frip_cut))
```

```{r}
print(c(
  "Using QC Cutoffs:",
  paste("QC FRAlt  : >",fralt_cut),
  paste("QC FRITSS : >",fritss_cut),
  paste("QC FRIP   : >",frip_cut)
))
```

[Return to Contents](#contents)

<a id="load_refs"></a>

#### Load Reference Datasets
```{r Load references}
# Altius Index
alt_idx_file <- system.file("reference/hg38_alt_dhs_index.bed.gz", package = "ATAComb")
stm(paste("Reading Altius Index from", alt_idx_file))
alt_idx_bed <- fread(alt_idx_file)

# Buenrostro Peak Reference
peaks_file <- system.file("reference/hg38_GSE123577_filtered.bed.gz", package = "ATAComb")
stm(paste("Reading Peak Reference from", peaks_file))
peaks_bed <- fread(peaks_file)

# Gene Body Reference
gene_bodies_file <- system.file("reference/hg38_ensemble93_gene_bodies.bed.gz", package = "ATAComb")
stm(paste("Reading ENSEMBL Gene Body Reference from", gene_bodies_file))
gene_bodies_bed <- fread(gene_bodies_file)

# TSS Reference
tss_file <- system.file("reference/hg38_ensemble93_tss_2kb.bed.gz",package = "ATAComb")
stm(paste("Reading ENSEMBL TSS Reference from", tss_file))
tss_bed <- fread(tss_file)
```

#### Load SampleSheet
```{r Load SampleSheet}
stm(paste0("Reading SampleSheet from ", in_key))

samplesheet <- read.csv(in_key)

in_sample <- as.character(samplesheet$SampleID[samplesheet$WellID == in_well])

out_prefix <- file.path(out_dir, paste0(in_batch, "_", in_sample, "_"))
```


[Return to Contents](#contents)

<a id="assemble_meta"></a>

### Generate Metadata

#### Read metadata sources
```{r}
stm(paste0("Reading 10x metadata from ", in_sin))

tenx_meta <- fread(in_sin)[-1,]
tenx_meta <- tenx_meta[, list(barcode, total, duplicate,mitochondrial , passed_filters)]
names(tenx_meta) <- c("original_barcodes","n_fragments","n_duplicate","n_mito","n_unique")

stm(paste0("Reading reference counts from ", in_pre))

alt_idx_counts <- fread(paste0(in_prefix, "_altius_total_counts.tsv.gz"), header = FALSE, col.names = c("original_barcodes","altius_count"))
peaks_counts <- fread(paste0(in_prefix, "_peaks_total_counts.tsv.gz"), header = FALSE, col.names = c("original_barcodes","peaks_count"))
tss_counts <- fread(paste0(in_prefix, "_tss_total_counts.tsv.gz"), header = FALSE, col.names = c("original_barcodes","tss_count"))
gene_bodies_counts <- fread(paste0(in_prefix, "_gene_bodies_total_counts.tsv.gz"), header = FALSE, col.names = c("original_barcodes","gene_bodies_count"))

meta <- tenx_meta[alt_idx_counts, on = "original_barcodes"]
meta <- meta[peaks_counts, on = "original_barcodes"]
meta <- meta[tss_counts, on = "original_barcodes"]
meta <- meta[gene_bodies_counts, on = "original_barcodes"]
```

#### Generate cell IDs
```{r Generate IDs}
stm("Generating Cell IDs")

meta$barcodes <- ids::uuid(nrow(meta),
                           drop_hyphens = TRUE, 
                           use_time = TRUE)
meta$cell_name <- ids::adjective_animal(nrow(meta),
                                        n_adjectives = 2,
                                        max_len = 10)
```

#### Add Well/Chip/Pool/Sample IDs
```{r}
stm("Adding Batch and Sample ID metadata")
meta$well_id <- in_well
meta$chip_id <- sub("W[0-9]+$","",meta$well_id)
meta$batch_id <- sub("-.+","",meta$chip_id)
meta$pbmc_sample_id <- in_sample
```

#### Filter metadata based on cutoffs
```{r}
stm("Filtering based on QC cutoffs")

meta$fralt <- meta$altius_count / meta$n_unique
meta$frip <- meta$peaks_count / meta$n_unique
meta$fritss <- meta$tss_count / meta$n_unique
filtered_meta <- meta[fralt > fralt_cut & 
                        frip > frip_cut &
                        fritss > fritss_cut,]
```

#### Write unfiltered and filtered metadata
```{r}
meta_out <- paste0(out_prefix, "all_metadata.csv.gz")
stm(paste("Writing unfiltered metadata to", meta_out))
fwrite(meta, meta_out)

filtered_meta_out <- paste0(out_prefix, "filtered_metadata.csv.gz")
stm(paste("Writing filtered metadata to", filtered_meta_out))
fwrite(filtered_meta, filtered_meta_out)
```

[Return to Contents](#contents)

<a id="mats"></a>

## Build Feature Matrices

#### Read overlaps with Altius index
```{r Altius overlaps}
stm("Building Altius Index matrix")

alt_dt <- fread(paste0(in_prefix, "_altius_sparse_matrix.tsv.gz"),
                header = FALSE,
                col.names = c("original_barcodes",
                              "i",
                              "x"),
                colClasses = list("character" = 1,
                                  "integer" = c(2,3)))

alt_dt <- alt_dt[original_barcodes %in% filtered_meta$original_barcodes]
```

#### Build Altius list
```{r Build Altius h5}
barcode_counts <- table(alt_dt$original_barcodes)
region_names <- paste(alt_idx_bed$V1,
                      alt_idx_bed$V2,
                      alt_idx_bed$V3,
                      sep = "_")
alt_list <- list(matrix = list(barcodes = names(barcode_counts),
                               data = alt_dt$x,
                               indices = alt_dt$i - 1,
                               indptr = c(0, unname(cumsum(barcode_counts))),
                               shape = c(nrow(alt_idx_bed),
                                         nrow(filtered_meta)),
                               features = list(name = region_names,
                                               id = alt_idx_bed$V4)))
rm(alt_dt)

alt_list$matrix$observations <- as.list(filtered_meta)
```

#### Write Altius Index .h5 file
```{r Write Altius h5}
alt_out_file <- paste0(out_prefix, "altius.h5")
stm(paste("Writing Altius Index h5 file to", alt_out_file))
write_h5_list(alt_list,
              alt_out_file)

rm(alt_list)
```

[Return to Contents](#contents)

#### Read overlaps with Peaks reference
```{r Peak overlaps}
stm("Building Peak matrix")

peaks_dt <- fread(paste0(in_prefix, "_peaks_sparse_matrix.tsv.gz"),
                  header = FALSE,
                  col.names = c("original_barcodes",
                                "i",
                                "x"),
                  colClasses = list("character" = 1,
                                    "integer" = c(2,3)))

peaks_dt <- peaks_dt[original_barcodes %in% filtered_meta$original_barcodes]
```

#### Build Peaks list
```{r Build Peaks h5}
barcode_counts <- table(peaks_dt$original_barcodes)
peak_names <- paste(peaks_bed$V1,
                    peaks_bed$V2,
                    peaks_bed$V3,
                    sep = "_")
peaks_list <- list(matrix = list(barcodes = names(barcode_counts),
                                 data = peaks_dt$x,
                                 indices = peaks_dt$i - 1,
                                 indptr = c(0, unname(cumsum(barcode_counts))),
                                 shape = c(length(peak_names),
                                           nrow(filtered_meta)),
                                 features = list(name = peak_names)))
rm(peaks_dt)

peaks_list$matrix$observations <- as.list(filtered_meta)
```

#### Write Peaks .h5 file
```{r Write Peaks h5}
peaks_out_file <- paste0(out_prefix, "peaks.h5")
stm(paste("Writing peaks h5 file to", peaks_out_file))
write_h5_list(peaks_list,
              peaks_out_file)

rm(peaks_list)
```

[Return to Contents](#contents)

#### Read overlaps with TSS reference
```{r TSS overlaps}
stm("Building TSS matrix")

tss_dt <- fread(paste0(in_prefix, "_tss_sparse_matrix.tsv.gz"),
                header = FALSE,
                col.names = c("original_barcodes",
                              "i",
                              "x"),
                colClasses = list("character" = 1,
                                  "integer" = c(2,3)))

tss_dt <- tss_dt[original_barcodes %in% filtered_meta$original_barcodes]
```

#### Build TSSs list
```{r Build TSS h5}
barcode_counts <- table(tss_dt$original_barcodes)
tss_names <- paste(tss_bed$V1,
                   tss_bed$V2,
                   tss_bed$V3,
                   sep = "_")
tss_list <- list(matrix = list(barcodes = names(barcode_counts),
                               data = tss_dt$x,
                               indices = tss_dt$i - 1,
                               indptr = c(0, cumsum(barcode_counts)),
                               shape = c(nrow(tss_bed),
                                         nrow(filtered_meta)),
                               features = list(name = tss_names)))
rm(tss_dt)

tss_list$matrix$observations <- as.list(filtered_meta)
```

#### Write TSSs .h5 file
```{r Write TSSs h5}
tss_out_file <- paste0(out_prefix, "tss.h5")
stm(paste("Writing tss h5 file to", tss_out_file))
write_h5_list(tss_list,
              tss_out_file)

rm(tss_list)
```

[Return to Contents](#contents)

#### Read overlaps with Gene Bodies reference
```{r Gene Bodies overlaps}
stm("Building Gene Bodies matrix")

gene_bodies_dt <- fread(paste0(in_prefix, "_gene_bodies_sparse_matrix.tsv.gz"),
                        header = FALSE,
                        col.names = c("original_barcodes",
                                      "i",
                                      "x"),
                        colClasses = list("character" = 1,
                                          "integer" = c(2,3)))

gene_bodies_dt <- gene_bodies_dt[original_barcodes %in% filtered_meta$original_barcodes]
```

#### Build Gene Bodies list
```{r Build Gene Bodies h5}
barcode_counts <- table(gene_bodies_dt$original_barcodes)
gene_bodies_names <- paste(gene_bodies_bed$V1,
                           gene_bodies_bed$V2,
                           gene_bodies_bed$V3,
                           sep = "_")
gene_bodies_list <- list(matrix = list(barcodes = names(barcode_counts),
                                       data = gene_bodies_dt$x,
                                       indices = gene_bodies_dt$i - 1,
                                       indptr = c(0, cumsum(barcode_counts)),
                                       shape = c(nrow(gene_bodies_bed),
                                                 nrow(filtered_meta)),
                                       features = list(name = gene_bodies_names)))
rm(gene_bodies_dt)

gene_bodies_list$matrix$observations <- as.list(filtered_meta)
```

Write Gene Bodies .h5 file
```{r Write Gene Bodies h5}
gene_bodies_out_file <- paste0(out_prefix, "gene_bodies.h5")
stm(paste("Writing gene_bodies h5 file to", gene_bodies_out_file))
write_h5_list(gene_bodies_list,
              gene_bodies_out_file)

rm(gene_bodies_list)
```

[Return to Contents](#contents)

### Window count matrices

#### Read 5k window counts
```{r 5k Window Matrices}
window_5k_chr <- read_chrom_sizes("hg38", window_size = 5e3)
window_5k_dt <- fread(paste0(in_prefix, "_window_5k_counts.tsv.gz"),
                      header = FALSE,
                      col.names = c("original_barcodes",
                                    "chr",
                                    "chr_i",
                                    "x"),
                      colClasses = list("character" = c(1,2),
                                        "integer" = c(3,4)))

window_5k_dt <- window_5k_dt[original_barcodes %in% filtered_meta$original_barcodes]
window_5k_offsets <- window_5k_chr[,c("chr","offset")]

window_5k_dt <- window_5k_dt[window_5k_offsets, on = "chr"]
window_5k_dt$i <- window_5k_dt$chr_i + window_5k_dt$offset
window_5k_dt <- window_5k_dt[!is.na(original_barcodes)]
window_5k_dt <- setorder(window_5k_dt, original_barcodes, i)
```

#### Build 5k window list
```{r Build 5k Window h5}
region_bed <- window_index_to_bed(1:sum(window_5k_chr$n_windows),
                                  window_5k_chr,
                                  window_size = 5e3)
region_names <- paste(region_bed$chr, region_bed$start, region_bed$end, sep = "_")
barcode_counts <- table(window_5k_dt$original_barcodes)
window_5k_list <- list(matrix = list(barcodes = names(barcode_counts),
                                     data = window_5k_dt$x,
                                     indices = window_5k_dt$i - 1,
                                     indptr = c(0, cumsum(barcode_counts)),
                                     shape = c(sum(window_5k_chr$n_windows),
                                               nrow(filtered_meta)),
                                     features = list(id = 1:length(region_bed),
                                                     name = region_names)))
rm(window_5k_dt)

window_5k_list$matrix$observations <- as.list(filtered_meta)
```

#### Write 5k Window .h5 file
```{r Write 5k Window h5}
window_5k_out_file <- paste0(out_prefix, "window_5k.h5")
stm(paste("Writing 5k Window h5 file to", window_5k_out_file))
write_h5_list(window_5k_list,
              window_5k_out_file)

rm(window_5k_list)
```

[Return to Contents](#contents)

#### Read 20k window counts
```{r 20k Window Matrices}
window_20k_offsets <- read_chrom_sizes("hg38", window_size = 2e4)
window_20k_dt <- fread(paste0(in_prefix, "_window_20k_counts.tsv.gz"),
                       header = FALSE,
                       col.names = c("original_barcodes",
                                     "chr",
                                     "chr_i",
                                     "x"),
                       colClasses = list("character" = c(1,2),
                                         "integer" = c(3,4)))

window_20k_dt <- window_20k_dt[original_barcodes %in% filtered_meta$original_barcodes]

window_20k_dt <- window_20k_dt[window_20k_offsets, on = "chr"]
window_20k_dt$i <- window_20k_dt$chr_i + window_20k_dt$offset
window_20k_dt <- window_20k_dt[!is.na(original_barcodes)]
window_20k_dt <- setorder(window_20k_dt, original_barcodes, i)
```

#### Build 20k window list
```{r Build 20k Window h5}
region_bed <- window_index_to_bed(1:sum(window_20k_offsets$n_windows),
                                  window_20k_offsets,
                                  window_size = 2e4)
region_names <- paste(region_bed$chr, region_bed$start, region_bed$end, sep = "_")
barcode_counts <- table(window_20k_dt$original_barcodes)
window_20k_list <- list(matrix = list(barcodes = names(barcode_counts),
                                      data = window_20k_dt$x,
                                      indices = window_20k_dt$i - 1,
                                      indptr = c(0, cumsum(barcode_counts)),
                                      shape = c(sum(window_20k_offsets$n_windows),
                                                nrow(filtered_meta)),
                                      features = list(name = region_names)))
rm(window_20k_dt)

window_20k_list$matrix$observations <- as.list(filtered_meta)
```

#### Write 20k window .h5 file
```{r Write 20k Window h5}
window_20k_out_file <- paste0(out_prefix, "window_20k.h5")
stm(paste("Writing 20k Window h5 file to", window_20k_out_file))
write_h5_list(window_20k_list,
              window_20k_out_file)

rm(window_20k_list)
```

[Return to Contents](#contents)

#### Read 100k window counts
```{r 100k Window Matrices}
window_100k_offsets <- read_chrom_sizes("hg38", window_size = 1e5)
window_100k_dt <- fread(paste0(in_prefix, "_window_100k_counts.tsv.gz"),
                        header = FALSE,
                        col.names = c("original_barcodes",
                                      "chr",
                                      "chr_i",
                                      "x"),
                        colClasses = list("character" = c(1,2),
                                          "integer" = c(3,4)))

window_100k_dt <- window_100k_dt[original_barcodes %in% filtered_meta$original_barcodes]

window_100k_dt <- window_100k_dt[window_100k_offsets, on = "chr"]
window_100k_dt$i <- window_100k_dt$chr_i + window_100k_dt$offset
window_100k_dt <- window_100k_dt[!is.na(original_barcodes)]
window_100k_dt <- setorder(window_100k_dt, original_barcodes, i)
```

#### Build 100k window list
```{r Build 100k Window h5}
region_bed <- window_index_to_bed(1:sum(window_100k_offsets$n_windows),
                                  window_100k_offsets,
                                  window_size = 1e5)
region_names <- paste(region_bed$chr, region_bed$start, region_bed$end, sep = "_")
barcode_counts <- table(window_100k_dt$original_barcodes)
window_100k_list <- list(matrix = list(barcodes = names(barcode_counts),
                                       data = window_100k_dt$x,
                                       indices = window_100k_dt$i - 1,
                                       indptr = c(0, cumsum(barcode_counts)),
                                       shape = c(sum(window_100k_offsets$n_windows),
                                                 nrow(filtered_meta)),
                                       features = list(name = region_names)))
rm(window_100k_dt)

window_100k_list$matrix$observations <- as.list(filtered_meta)
```

#### Write 100k window .h5 file
```{r Write 100k Window h5}
window_100k_out_file <- paste0(out_prefix, "window_100k.h5")
stm(paste("Writing 100k Window h5 file to", window_100k_out_file))
write_h5_list(window_100k_list,
              window_100k_out_file)

rm(window_100k_list)
```

[Return to Contents](#contents)

<a id="qc_stats"></a>

### QC Stats

```{r}
qc_list <- list(report_type = "atac_qc",
                report_datetime = as.character(start_time),
                report_uuid = ids::uuid(use_time = TRUE),
                package = "ATAComb",
                package_version = sessionInfo()$otherPkgs$ATAComb$Version,
                well_id = in_well)

out_json <- paste0(out_prefix, "atac_qc_metrics.json")
```

[Return to Contents](#contents)

<a id="barcode_stats"></a>

#### Barcode QC Stats
```{r}
barcode_conditions <- list("all" = rep(TRUE, nrow(meta)),
                           "pass_qc" = meta$fralt > fralt_cut & meta$frip > frip_cut & meta$fritss > fritss_cut,
                           "fail_qc" = meta$fralt < fralt_cut | meta$frip < frip_cut | meta$fritss < fritss_cut,
                           "fail_fralt" = meta$fralt < fralt_cut,
                           "fail_frip" = meta$frip < frip_cut,
                           "fail_fritss" = meta$fritss < fritss_cut,
                           "fail_all" = meta$fralt < fralt_cut & meta$frip < frip_cut & meta$fritss < fritss_cut)

perc_of <- function(x, d) {
  round(sum(x) / d * 100, 2)
}

q_of <- function(x, target, q) {
  vals <- target[x]
  round(quantile(vals, probs = q), 4)
}

barcode_qc_stats <- data.frame(metric = c("Barcodes with > 1k reads",
                                          "Barcodes pass QC",
                                          "Barcodes fail QC",
                                          "Barcodes fail FRAlt",
                                          "Barcodes fail FRIP",
                                          "Barcodes fail FRITSS",
                                          "Barcodes fail all"),
                               count = sapply(barcode_conditions, sum),
                               percent = sapply(barcode_conditions, perc_of, nrow(meta)),
                               median_fragments = sapply(barcode_conditions, q_of, meta$n_fragments, .50),
                               median_unique = sapply(barcode_conditions, q_of, meta$n_unique, .50),
                               median_fralt = sapply(barcode_conditions, q_of, meta$fralt, .50),
                               median_frip = sapply(barcode_conditions, q_of, meta$frip, .50),
                               median_fritss = sapply(barcode_conditions, q_of, meta$fritss, .50))

qc_list$barcode_qc_stats <- as.list(barcode_qc_stats)
qc_list$barcode_qc_stats$metric <- c("barcodes_gt1k","barcodes_pass_qc","barcodes_fail_qc",
                                     "barcodes_fail_fralt","barcodes_fail_frip","barcodes_fail_fritss","barcodes_fail_all")

qc_table(barcode_qc_stats)
```

[Return to Contents](#contents)

<a id="fragment_stats"></a>

#### Fragment QC stats
```{r}
metrics_summary <- read.csv(in_sum)

saturation_counts <- fread(paste0(in_prefix, "_saturation_curve.tsv.gz"),
                           header = F,
                           col.names = c("count", "frequency"))
saturation_counts$count <- as.numeric(sub("\\r","",saturation_counts$count))

fragment_qc_stats <- data.frame(metric = c("Sequenced Fragments",
                                           "Mitochondrial Fragments",
                                           "Total in Barcodes with > 1k Reads",
                                           "Total in Barcodes passing QC",
                                           "Total in Barcodes failing QC",
                                           "Total Unique",
                                           "Unique in Barcodes with > 1k Reads",
                                           "Unique in Barcodes passing QC",
                                           "Unique in Barcodes failing QC"),
                                count = c(metrics_summary$num_fragments,
                                          ceiling(metrics_summary$num_fragments * metrics_summary$frac_waste_mitochondrial),
                                          sum(meta$n_fragments),
                                          sum(filtered_meta$n_fragments),
                                          sum(meta$n_fragments) - sum(filtered_meta$n_fragments),
                                          sum(saturation_counts$frequency),
                                          sum(meta$n_unique),
                                          sum(filtered_meta$n_unique),
                                          sum(meta$n_unique) - sum(filtered_meta$n_unique)))
fragment_qc_stats$millions <- round(fragment_qc_stats$count / 1e6, 2)
fragment_qc_stats$percent_of_sequenced <- round(fragment_qc_stats$count / fragment_qc_stats$count[1] * 100, 2)
fragment_qc_stats$percent_of_unique <- round(fragment_qc_stats$count / sum(saturation_counts$frequency) * 100, 2)
fragment_qc_stats$percent_of_unique[1:5] <- NA

qc_list$fragment_qc_stats <- as.list(fragment_qc_stats)
qc_list$fragment_qc_stats$metric <- c("total_frags","mito_frags",
                                      "total_frags_gt1k","total_frags_pass_qc","total_frags_fail_qc",
                                      "unique_frags","unique_frags_gt1k","unique_frags_pass_qc","unique_frags_fail_qc")

qc_table(fragment_qc_stats)
```

[Return to Contents](#contents)

<a id="saturation_stats"></a>

#### Saturation metrics
```{r}
stm("Generating library diversity projection")

saturation_mat <- as.matrix(as.data.frame(saturation_counts))

total_metrics <- data.frame(total_reads = metrics_summary$num_fragments,
                            total_umis = metrics_summary$total_usable_fragments,
                            total_counts = sum(saturation_mat[,"count"] * saturation_mat[,"frequency"]))

saturation_projection <- suppressWarnings(
  diversity_projection(saturation_mat,
                       total_metrics,
                       max_val = 2e9)
)

saturation_projection$type <- ifelse(saturation_projection$n_raw_reads < total_metrics$total_reads,
                                     "actual",
                                     "projected")

saturation_projection$ratio <- saturation_projection$n_raw_reads/saturation_projection$expected_umis
saturation_projection$ratio[1] <- 0

# Signal UMIs: UMIs in cells passing QC in peaks
saturation_projection$signal_umis <- floor(saturation_projection$expected_umis * 
                                             fragment_qc_stats$percent_of_unique[fragment_qc_stats$metric == "Unique in Barcodes passing QC"] / 100 *
                                             sum(filtered_meta$peaks_count) / sum(filtered_meta$n_unique))

saturation_projection$signal_ratio <- saturation_projection$n_raw_reads / saturation_projection$signal_umis
saturation_projection$signal_ratio[1] <- 0

# Precomputing millions of reads for display
saturation_projection$M_raw_reads <- round(saturation_projection$n_raw_reads/1e6, 2)
saturation_projection$M_umis <- round(saturation_projection$expected_umis/1e6, 2)
saturation_projection$M_signal_umis <- round(saturation_projection$signal_umis/1e6,2)
```

#### Output projection values
```{r}
out_proj <- paste0(out_prefix, "saturation_projection.csv.gz")
stm(paste0("OUT Sat. Projection   : ", out_proj))
print(paste0("OUT Sat. Projection    : ", out_proj))
stm(paste("Writing Saturation Projection to", out_proj))

fwrite(saturation_projection,
       out_proj)
```

#### Compute Break-points for specific ratios of raw reads to UMIs or signal UMIs
```{r}
# 2:1, 3:1, and 4:1 break points for total UMIs
saturation_break_idx <- sapply(2:4, function(x) which.min(abs(saturation_projection$ratio-x)))

saturation_breaks <- saturation_projection[saturation_break_idx,]
saturation_breaks$label <- c("2:1","3:1","4:1")
saturation_breaks$value <- c(2:4)

# Filter out breaks that are nearest, but not very close to the actual break value
# removes things like adding a break at the final point even if it doesn't reach 4:1
saturation_breaks <- saturation_breaks[abs(saturation_breaks$value - saturation_breaks$ratio) < 0.1,]

# 4:1, 8:1, and 16:1 break points for signal UMIs
signal_break_idx <- sapply(c(4,8,16), function(x) which.min(abs(saturation_projection$signal_ratio-x)))

signal_breaks <- saturation_projection[signal_break_idx,]
signal_breaks$label <- c("2:1","4:1","16:1")
signal_breaks$value <- c(4,8,16)

signal_breaks <- signal_breaks[abs(signal_breaks$value - signal_breaks$signal_ratio) < 0.1,]

```

#### Display Break points for saturation of UMIs
```{r}
# Abbreviated table for display
saturation_qc <- data.frame(break_target = saturation_breaks$value,
                            umi_ratio = round(saturation_breaks$ratio,2),
                            M_raw_reads = saturation_breaks$M_raw_reads,
                            M_umis = saturation_breaks$M_umis,
                            M_signal_umis = saturation_breaks$M_signal_umis)

qc_list$saturation <- list()
qc_list$saturation$umi_breaks <- as.list(saturation_qc)

qc_table(saturation_qc)
```

#### Display Break points for saturation of signal UMIs
```{r}
signal_qc <- data.frame(break_target = signal_breaks$value,
                        signal_ratio = round(signal_breaks$signal_ratio, 2),
                        M_raw_reads = signal_breaks$M_raw_reads,
                        M_umis = signal_breaks$M_umis,
                        M_signal_umis = signal_breaks$M_signal_umis)

qc_list$saturation$signal_breaks <- as.list(signal_qc)

qc_table(signal_qc)
```

#### Saturation summary by sequencing depth
```{r}
regular_break_idx <- which(saturation_projection$M_raw_reads %in% seq(200, 2000, by = 200))

regular_qc <- data.frame(umi_ratio = round(saturation_projection$ratio[regular_break_idx], 2),
                         signal_ratio = round(saturation_projection$signal_ratio[regular_break_idx], 2),
                         M_raw_reads = saturation_projection$M_raw_reads[regular_break_idx],
                         M_umis = saturation_projection$M_umis[regular_break_idx],
                         M_signal_umis = saturation_projection$M_signal_umis[regular_break_idx])

qc_list$saturation$raw_breaks <- as.list(regular_qc)

qc_table(regular_qc)
```

[Return to Contents](#contents)

<a id="tenx_stats"></a>

#### 10x Metrics
```{r}
tenx_metrics <- data.frame(metric = names(metrics_summary),
                           values = round(unlist(as.numeric(metrics_summary[1,])), 4))

qc_list$tenx_metrics <- as.list(metrics_summary)

qc_table(tenx_metrics)
```

[Return to Contents](#contents)

<a id="qc_plots"></a>

### QC Plots

<a id="saturation_plot"></a>

#### Plot Library Saturation
```{r}
reference_projections <- fread(system.file("reference/saturation/reference_projections.csv",
                                           package = "ATAComb"))
reference_projections$dataset <- paste0("ref_",reference_projections$dataset)

umi_saturation_plot <- ggplot() +
  geom_line(data = reference_projections,
            aes(x = n_raw_reads,
                y = expected_umis,
                group = dataset,
                color = dataset)) +
  geom_line(data = saturation_projection,
            aes(x = n_raw_reads,
                y = expected_umis,
                group = type,
                color = type),
            size = 2)  +
  scale_color_brewer("Data Type",
                     type = "qual",
                     palette = 2) +
  scale_x_continuous("N Raw Reads (millions)",
                     expand = c(0,0),
                     breaks = seq(0, 2e9, by = 2.5e8),
                     labels = seq(0, 2000, by = 250)) +
  scale_y_continuous("N Unique Fragments (millions)", 
                     expand = c(0,0),
                     limits = c(0, 7.5e8),
                     breaks = seq(0, 7.5e8, by = 2.5e8),
                     labels = seq(0, 750, by = 250)) +
  theme_bw() +
  theme(panel.grid.minor = element_blank())

if(nrow(saturation_breaks) > 0) {
  umi_saturation_plot <- umi_saturation_plot +
    geom_segment(data = saturation_breaks,
                 aes(x = 0, xend = n_raw_reads,
                     y = expected_umis, yend = expected_umis),
                 linetype="dashed") +
    geom_segment(data = saturation_breaks,
                 aes(x = n_raw_reads, xend = n_raw_reads,
                     y = 0, yend = expected_umis),
                 linetype="dashed") +
    geom_text(data = saturation_breaks,
              aes(x = n_raw_reads, y = expected_umis + 1e7,
                  label = label),
              hjust = 0,
              vjust = 0)
}

umi_saturation_plot
```

[Return to Contents](#contents)

<a id="signal_plot"></a>

#### Plot saturation of signal UMIs
```{r}
signal_saturation_plot <- ggplot() +
  geom_line(data = reference_projections,
            aes(x = n_raw_reads,
                y = signal_umis,
                group = dataset,
                color = dataset)) +
  geom_line(data = saturation_projection,
            aes(x = n_raw_reads,
                y = signal_umis,
                group = type,
                color = type),
            size = 2) +
  scale_color_brewer("Data Type",
                     type = "qual",
                     palette = 2) +
  scale_x_continuous("N Raw Reads (millions)",
                     expand = c(0,0),
                     breaks = seq(0, 2e9, by = 2.5e8),
                     labels = seq(0, 2000, by = 250)) +
  scale_y_continuous("N Unique Fragments in Peaks (millions)", 
                     expand = c(0,0),
                     limits = c(0, 2e8),
                     breaks = seq(0, 2e8, by = 2.5e7),
                     labels = seq(0, 200, by = 25)) +
  theme_bw() +
  theme(panel.grid.minor = element_blank())

if(nrow(signal_breaks) > 0) {
  signal_saturation_plot <- signal_saturation_plot +
    geom_segment(data = signal_breaks,
                 aes(x = 0, xend = n_raw_reads,
                     y = signal_umis, yend = signal_umis),
                 linetype="dashed") +
    geom_segment(data = signal_breaks,
                 aes(x = n_raw_reads, xend = n_raw_reads,
                     y = 0, yend = signal_umis),
                 linetype="dashed") +
    geom_text(data = signal_breaks,
              aes(x = n_raw_reads, y = signal_umis + 5e6,
                  label = label),
              vjust = 0)
}

signal_saturation_plot
```

[Return to Contents](#contents)

<a id="width_plot"></a>

#### Plot Fragment Widths
```{r}
frag_widths <- fread(paste0(in_prefix, "_fragment_widths.tsv.gz"),
                     header = FALSE,
                     col.names = c("original_barcodes", "width", "count"),
                     colClasses = list("character" = 1,
                                       "integer" = c(2,3)))
frag_widths$pass_fail <- ifelse(frag_widths$original_barcodes %in% filtered_meta$original_barcodes,
                                "pass_qc",
                                "fail_qc")
frag_widths <- frag_widths[, sum(count), by = list(pass_fail, width)]
names(frag_widths)[3] <- "totals"
frag_widths <- frag_widths[, frac:=totals/sum(totals), by = list(pass_fail)]
frag_widths <- frag_widths[width <= 750]

ggplot(data = frag_widths) +
  geom_line(aes(x = width,
                y = frac,
                group = pass_fail,
                color = pass_fail),
            size = 1) +
  xlim(0, 750) +
  facet_grid(rows = vars(pass_fail)) +
  theme_bw()
```

[Return to Contents](#contents)

<a id="all_hist_plot"></a>

#### Histograms for all cells
```{r}
n_reads_hist <- qc_hist_plot(meta,
                             column = "n_unique",
                             name_x = "N Unique Fragments per Cell",
                             log_x = TRUE,
                             fill = "dodgerblue",
                             target = 2500,
                             y_max = 1000)
fralt_hist <- qc_frac_hist_plot(meta,
                                column = "fralt",
                                name_x = "Frac Reads in Altius Index (FRAlt)",
                                fill = "darkgreen",
                                target = fralt_cut,
                                y_max = 8000)
frip_hist <- qc_frac_hist_plot(meta,
                               column = "frip",
                               name_x = "Frac Reads in Peaks (FRIP)",
                               fill = "mediumorchid3",
                               target = frip_cut,
                               y_max = 5000)
fritss_hist <- qc_frac_hist_plot(meta,
                                 column = "fritss",
                                 name_x = "Frac Reads in TSS (FRITSS)",
                                 fill = "orangered",
                                 target = fritss_cut,
                                 y_max = 5000)
plot_list <- list(n_reads_hist,
                  fralt_hist,
                  frip_hist,
                  fritss_hist)

cowplot::plot_grid(plotlist = plot_list,
                   nrow = 4, ncol = 1)
```

[Return to Contents](#contents)

<a id="reads_fralt_plot"></a>

#### Reads vs FRAlt scatter
```{r}
qc_scatter_plot(meta,
                column_x = "n_unique",
                name_x = "N Unique Fragments per Cell",
                column_y = "fralt",
                name_y = "Frac Fragments in Altius Index (FRAlt)",
                log_x = TRUE, log_y = FALSE, frac_y = TRUE,
                show_targets = FALSE,
                color = "darkgreen") +
  geom_vline(aes(xintercept = 2.5e3), linetype = "dashed", size = 0.2) +
  geom_hline(aes(yintercept = fralt_cut), linetype = "dashed", size = 0.2)
```


[Return to Contents](#contents)

<a id="reads_frip_plot"></a>

#### Reads vs FRIP scatter
```{r}
qc_scatter_plot(meta,
                column_x = "n_unique",
                name_x = "N Unique Fragments per Cell",
                column_y = "frip",
                name_y = "Frac Fragments in Peaks (FRIP)",
                log_x = TRUE, log_y = FALSE, frac_y = TRUE,
                show_targets = FALSE,
                color = "darkgreen") +
  geom_vline(aes(xintercept = 2.5e3), linetype = "dashed", size = 0.2) +
  geom_hline(aes(yintercept = frip_cut), linetype = "dashed", size = 0.2)
```


[Return to Contents](#contents)

<a id="frip_fritss_plot"></a>

#### FRIP vs FRITSS scatter
```{r}
qc_scatter_plot(meta,
                column_x = "frip",
                name_x = "Frac Reads in Peaks (FRIP)",
                column_y = "fritss",
                name_y = "Frac Reads in TSS (FRITSS)",
                log_x = FALSE, frac_x = TRUE, log_y = FALSE, frac_y = TRUE,
                show_targets = FALSE,
                color = "mediumorchid3") +
  geom_vline(aes(xintercept = fritss_cut), linetype = "dashed", size = 0.2) +
  geom_hline(aes(yintercept = frip_cut), linetype = "dashed", size = 0.2)
```


[Return to Contents](#contents)

<a id="rip_ritss_plot"></a>

#### RIP vs RITSS scatter
```{r}
qc_scatter_plot(meta,
                column_x = "peaks_count",
                name_x = "N Reads in Peaks (FRIP)",
                column_y = "tss_count",
                name_y = "N Reads in TSS (FRITSS)",
                log_x = TRUE, frac_x = FALSE, log_y = TRUE, frac_y = FALSE,
                show_targets = TRUE,
                color = "orangered")
```

[Return to Contents](#contents)

<a id="filtered_hist_plot"></a>

#### Filtered QC Histograms
```{r}
n_reads_hist <- qc_hist_plot(filtered_meta,
                             column = "n_unique",
                             name_x = "N Unique Fragments per Cell",
                             log_x = TRUE,
                             fill = "dodgerblue",
                             target = 2500,
                             y_max = 1000)
fralt_hist <- qc_frac_hist_plot(filtered_meta,
                                column = "fralt",
                                name_x = "Frac Reads in Altius Index (FRAlt)",
                                fill = "darkgreen",
                                target = fralt_cut,
                                y_max = 8000)
frip_hist <- qc_frac_hist_plot(filtered_meta,
                               column = "frip",
                               name_x = "Frac Reads in Peaks (FRIP)",
                               fill = "mediumorchid3",
                               target = frip_cut,
                               y_max = 5000)
fritss_hist <- qc_frac_hist_plot(filtered_meta,
                                 column = "fritss",
                                 name_x = "Frac Reads in TSS (FRITSS)",
                                 fill = "orangered",
                                 target = fritss_cut,
                                 y_max = 5000)
plot_list <- list(n_reads_hist,
                  fralt_hist,
                  frip_hist,
                  fritss_hist)

cowplot::plot_grid(plotlist = plot_list,
                   nrow = 4, ncol = 1)
```

[Return to Contents](#contents)

<a id="json_out"></a>

### Write QC JSON

```{r Save QC JSON}
stm(paste0("Writing JSON to",out_json))

qc_list_json <- jsonlite::toJSON(qc_list,
                                 auto_unbox = TRUE,
                                 pretty = TRUE)

writeLines(qc_list_json,
           out_json)
```

[Return to Contents](#contents)

<a id="session_info"></a>

## Session Information

```{r Session Info}
sessionInfo()
```

Total time elapsed
```{r Show Time}
end_time <- Sys.time()
diff_time <- end_time - start_time
time_message <- paste0("Elapsed Time: ", 
                       round(diff_time, 3),
                       " ", units(diff_time))
print(time_message)
stm(time_message)
stm("H5 metadata process complete.")
```

[Return to Contents](#contents)
