#' Read 10x scATAC-seq fragment data
#'
#' @param outs_dir An outs/ directory generated by cellranger-atac count. Must contain fragments.tsv.gz and singlecell.csv
#' @param min_reads A numeric value indicating the minimum number of reads required to retain a cell barcode
#' @param verbose A logical value indicating whether or not to display messages.
#'
#' @return a list of GenomicRanges objects, one per cell barcode
#' @export
read_10x_fragments <- function(outs_dir,
                               min_reads = 1000,
                               verbose = TRUE) {

  fragments_file <- file.path(outs_dir, "fragments.tsv.gz")
  if(!file.exists(fragments_file)) {stop(paste("Can't find fragments.tsv.gz in", outs_dir))}

  singlecell_file <- file.path(outs_dir, "singlecell.csv")
  if(!file.exists(singlecell_file)) {stop(paste("Can't find singlecell.csv in", outs_dir))}

  if(verbose) {cat("reading singlecell.csv\n")}
  singlecell <- data.table::fread(singlecell_file)
  singlecell <- singlecell[-1,]
  if(verbose) {cat("filtering singlecell.csv\n")}
  singlecell <- singlecell[passed_filters >= min_reads,]

  if(verbose) {cat("reading fragments.tsv.gz\n")}
  fragments <- vroom::vroom(fragments_file,
                            delim = "\t",
                            col_names = c("chr","start","end","barcode","n_reads"),
                            col_types = c(chr = "c", start = "i", end = "i", barcode = "c", n_reads = "-"))

  if(verbose) {cat("filtering fragments.tsv.gz\n")}
  fragments <- fragments[fragments$barcode %in% singlecell$barcode,]

  if(verbose) {cat("splitting fragments by barcode\n")}
  split_fragments <- split(fragments, fragments$barcode)

  return(split_fragments)
}


#' Convert a list of fragment data.table to GenomicRanges
#'
#' @param fragments A list of data.tables generated by read_10x_fragments()
#' @param n_threads A numeric value indicating how many threads to use for parallelization. Only works on UNIX-like systems. Default is 1.
#'
#' @return a list of GRanges objects
#' @export
convert_fragments_gr <- function(fragments,
                                 n_threads = 1) {
  if(n_threads == 1) {
    lapply(fragments,
           function(x) {
             GenomicRanges::GRanges(seqnames = x[["V1"]],
                                    IRanges::IRanges(start = x[["V2"]],
                                                     end = x[["V3"]]))
           })
  } else {
    starting_order <- names(fragments)
    res <- parallel::mclapply(fragments,
                              function(x) {
                                GenomicRanges::GRanges(seqnames = x$V1,
                                                       IRanges::IRanges(start = x$V2,
                                                                        end = x$V3))
                              },
                              mc.cores = n_threads)
    res <- res[starting_order]
    return(res)
  }
}
